Add tag-level profile scoping. (#361)
ietf-rats-wg/draft-ietf-rats-corim: draft-ietf-rats-corim.md

For an example profile definition, see {{-psa-endorsements}}.

> A tag (e.g., CoMID, CoBOM) MAY specify a profile identifier to take precedence for the interpretation of the remainder of the tag contents if the following requirements hold:

Reviewer: ```suggestion
A tag (e.g., CoMID) MAY specify a profile identifier to take precedence for the interpretation of the remainder of the tag contents if the following requirements hold:
```

*  A profile's interpretation MUST be consistent regardless of where the profile identifier is supplied; whether in `corim-map.profile` or in a tag.
*  The `corim-map.profile` scope applies to all tags within the `corim-map` unless specifically overidden by a profile in a tag-specific profile.

> 

Author: ```suggestion

The scope of the profile identifier in a tag does not change the interpretation of any cross-referenced tags or CoRIMs.
```

For an example profile definition, see {{-psa-endorsements}}.

A tag (e.g., CoMID, CoBOM) MAY specify a profile identifier to take precedence for the interpretation of the remainder of the tag contents if the following requirements hold:

> *  Any interpretation of the tag that depends on contents from the `corim-map` containing the tag have the same meaning across the tag's profile and the `corim-map.profile`.

Reviewer 2: ```suggestion
*  A profile's interpretation MUST be consistent regardless of where it is supplied; whether in `corim-map.profile` or in a tag.
```

*  The `corim-map.profile` permits (exclusively or by omission) the tag to to use the profile.

> A profile MUST NOT have a different interpretation when specified on a tag; the tag should behave as if it were in a CoRIM with the profile specified in the `corim-map` instead.
> 

Reviewer 2: ```suggestion

```
The reworded line 385 seems to be inclusive of intent of this statement.

For an example profile definition, see {{-psa-endorsements}}.

A tag (e.g., CoMID, CoBOM) MAY specify a profile identifier to take precedence for the interpretation of the remainder of the tag contents if the following requirements hold:

*  Any interpretation of the tag that depends on contents from the `corim-map` containing the tag have the same meaning across the tag's profile and the `corim-map.profile`.
*  The `corim-map.profile` permits (exclusively or by omission) the tag to to use the profile.

> 

Reviewer: Also worth specifying explicitly that the scope of the profile defined at the tag level is ONLY restricted to the tag where it is defined. If no profile is defined in the tag then the CoRIM profile should be used for understanding the semantics

Author: "for the interpretation of the remainder of the tag contents if the following requirements hold:"

I think this is handled?

  modules.
  Described in {{sec-comid-triples}}.

> * `profile-override` (index 5): A profile identifier that replaces the

Reviewer 2: ```suggestion
* `profile-override` (index 5): A profile identifier that overrides the
```

  `corim-map`'s profile to interpret the contents of the `concise-mid-tag`. The
  `corim-map`'s profile MUST NOT change the interpretation of this `CoMID` when

>   the `profile-override` is specified.

Reviewer 2: ```suggestion
  a `profile-override` value is supplied.
```

> * `profile-override` (index 5): A profile identifier that overrides the
>   `corim-map.profile` to interpret the contents of the `concise-mid-tag`. 
> The `corim-map.profile` MUST NOT change the interpretation of the `CoMID` tag when
>   a `profile-override` value is supplied.

Author: ```suggestion
* `profile` (index 5): A profile identifier that overrides the
  `corim-map.profile` to interpret the contents of the `concise-mid-tag`. 
The `corim-map.profile` MUST NOT change the interpretation of the `CoMID` tag when
  a `profile` value is supplied.
```

CDDL map extensibility enables a CoRIM profile to extend the base CoRIM CDDL definition.
CDDL map extension points have the form `($$NAME-extension)` where "NAME" is the name of the map and '$$' signifies map extensibility.
Typically, map extension requires a convention for code point naming that avoids code-point reuse.

> Well-known code points may be in a registry, such as CoSWID {{-coswid}}.

Author: No, Ned removed it in ddceeb4219e2e1f8dfccca8b48a2e693860227d1...

> Well-known code points may be in a registry, such as CoSWID {{-coswid}}.

Reviewer 3: re-reading it in context, ISTM that L1358 can be dropped altogether.

> Well-known code points may be in a registry, such as CoSWID {{-coswid}}.

Author: I think we need to be clear that we're not suggesting that a profile is free to extend all extension points freely with negative values when the data type is under control of another RFC/registry.

> Well-known code points may be in a registry, such as CoSWID {{-coswid}}.

Reviewer 2: > re-reading it in context, ISTM that L1358 can be dropped altogether.

The tooling was failing because the same reference was duplicated. Somehow the tooling equated them as the same reference. I removed one to make it build successfully.

* `bom-validity` (index 2): Specifies the validity period of the CoBOM.
  Described in {{sec-common-validity}}.

* `profile` (index 3): An optional profile identifier that overrides the
  `corim-map.profile` to interpret the contents of the `concise-bom-tag`.

> The `corim-map.profile`  MUST NOT change the interpretation of the `CoBOM` tag when
>   a `profile` value is supplied.

Reviewer 3: ditto:

```suggestion
The CoBOM profile identifier takes precedence over the CoRIM profile identifier.
```

* `bom-validity` (index 2): Specifies the validity period of the CoBOM.
  Described in {{sec-common-validity}}.

> * `profile-override` (index 3): A profile identifier that replaces the

Reviewer 2: ```suggestion
* `profile-override` (index 3): A profile identifier that overrides the
```

  `corim-map`'s profile to interpret the contents of the `concise-bom-tag`. The
  `corim-map`'s profile MUST NOT change the interpretation of this `CoBOM` when

>   the `profile-override` is specified

Reviewer 2: ```suggestion
  a `profile-override` value is supplied.
```

> * `profile-override` (index 3): A profile identifier that overrides the
>   `corim-map.profile` to interpret the contents of the `concise-bom-tag`.
> The `corim-map.profile`  MUST NOT change the interpretation of the `CoBOM` tag when
>   a `profile-override` value is supplied.

Author: ```suggestion
* `profile` (index 3): A profile identifier that overrides the
  `corim-map.profile` to interpret the contents of the `concise-bom-tag`.
The `corim-map.profile`  MUST NOT change the interpretation of the `CoBOM` tag when
  a `profile` value is supplied.
```

* `bom-validity` (index 2): Specifies the validity period of the CoBOM.
  Described in {{sec-common-validity}}.

> * `profile` (index 3): An optional profile identifier that overrides the

Reviewer 3: Since CoBOMs contain pointers to other tags each potentially sporting its own profile, is there a "homogeneity" requirement?  I.e., that the pointed tags have the same profile?

Author: No, the scope is to the lexical structure itself. It does not have "dynamic scope" semantics. Added a line.

* `bom-validity` (index 2): Specifies the validity period of the CoBOM.
  Described in {{sec-common-validity}}.

* `profile` (index 3): An optional profile identifier that overrides the
  `corim-map.profile` to interpret the contents of the `concise-bom-tag`.

> The CoBOM profile identifier takes precedence over the CoRIM profile identifier.

Author: ```suggestion
The CoBOM profile identifier takes precedence over the CoRIM profile identifier.
The CoBOM profile identifier does not override the profile of any cross-referenced tag.
```

Note that a Verifier may compare Reference Values in any order, so the comparison should not be stateful.

## Concise Software Identifier Extension {#sec-coswid}

> A CoRIM may contain a CoSWID tag as specified in ({{sec-corim-tags}}).

Reviewer 3: ```suggestion
A CoRIM may contain a CoSWID tag as specified in {{sec-corim-tags}}.
```

> This specification registers an extension to `concise-swid-tag` using the `$$coswid-extension` mechanism. This extension defines a CoSWID profile code point that allows for an optional CoSWID profile identifier.

Reviewer 3: ```suggestion
This specification registers an extension to `concise-swid-tag` using the `$$coswid-extension` socket.
This extension defines a CoSWID profile code point that allows for an optional CoSWID profile identifier.
```

> The CoSWID profile identifier takes precedence over the CoRIM profile identifier.

Reviewer 3: I like this.  We can reuse it for CoMID and CoBOM.

* `profile` (index 58): A profile identifier that overrides the `corim-map.profile` to interpret the contents of the `concise-swid-tag`.

> The `corim-map.profile` MUST NOT change the interpretation of the `CoSWID` tag when a `profile` value is supplied.

Reviewer 3: I think L2506 is sufficient.
```suggestion
```

Note that a Verifier may compare Reference Values in any order, so the comparison should not be stateful.

## Concise Software Identifier Extension {#sec-coswid}

A CoRIM may contain a CoSWID tag as specified in {{-coswid}}.

> The CoRIM profile allows for the specifier to provide meaning to extension points in the "Specification Required" range of the extension points `$version-scheme` and `$$coswid-extension`.

Reviewer 2: Reword this sentence. I find it confusing. Perhaps it recites CoSWID rules for using extension points. If so, it can be assumed that this spec will follow IANA / CoSWID rules for extension points.

> The CoRIM profile allows for the specifier to provide meaning to extension points in the "Specification Required" range of the extension points `$version-scheme` and `$$coswid-extension`.

Author: ```suggestion
```
Is removal okay since I already describe essentially the same thing in the profile section above?

> The CoRIM profile is not permitted to specify negative values for these extension points.

Reviewer 2: ```suggestion

```
Not needed as this spec will follow CoSWID requirements for extension.

> To allow for tag-level scoping of profile identifier to bundle multiple items under a single CoRIM, this specification registers a `$$coswid-extension` to specifically name the profile that gives meaning to the profile-defined extensions.

Reviewer 2: ```suggestion
This specification registers an extension to `concise-swid-tag` using the `$$coswid-extension` mechanism. This extension defines a CoSWID profile code point that allows for an optional CoSWID profile identifier.
The CoSWID profile identifier takes precedence over the CoRIM profile identifier.

* `profile-override` (index 58): A profile identifier that overrides the `corim-map.profile` to interpret the contents of the `concise-swid-tag`.
The `corim-map.profile` MUST NOT change the interpretation of the `CoSWID` tag when a `profile-override` value is supplied.
```

Author: Since this registration is to CoSWID Items, I specifically avoided "profile-override". A CoSWID could be created on its own with the "profile" code point as a reference in itself to the specification for any extension on the base spec.

Reviewer 2: I kind of think "profile" is better overall since the spec language describes scoping semantics normatively. Other specs could incorporate CoMID or CoBOM subsets resulting in a similar dynamic as with CoSWID.

Author: I shall update the other occurrences as well then.

| application/corim-unsigned+cbor | - | TBD2 | {{&SELF}} |
{: align="left" title="New Content-Formats"}

## CoSWID Item Registration

> IANA is requested to add the following CoSWID Items to the "Concise Software Identifier"
> registry {{!IANA.coswid}}.

Reviewer 3: ```suggestion
IANA is requested to add the following CoSWID Items to the "CoSWID Items" registry within the "Concise Software Identifier" registry group {{-coswid-reg}}.
```
